solidity/contracts/hooks/CctpHook.sol

```diff
@@ -0,0 +1,130 @@
+    // see https://developers.circle.com/stablecoins/transfer-usdc-on-testnet-from-ethereum-to-avalanche
+    uint32 internal constant CCTP_V2_DEFAULT_MIN_FINALITY_THRESHOLD = 2000;
+
+    // we keep the token messenger here in order to
+    // provide the right quote
+    ITokenMessenger immutable tokenMessenger;
+    IPostDispatchHook public immutable childHook;
+
+    // we specify the IGP here in order to make
+    // the transfer `user => TokenBridge => CctpHook`
+    // easier. A StaticAggregationHook has been considered
+    // but there is no consistent way to retrieve the
+    // correct hook when performing the approval on the
+    // token bridge
+    IPostDispatchHook igp;
+
+    // ============ Constructor ============
+    constructor(
+        ITokenMessenger _tokenMessenger,
+        IPostDispatchHook _igp,
+        IPostDispatchHook _childHook
+    ) {
+        igp = _igp;
+        tokenMessenger = _tokenMessenger;
+        childHook = _childHook;
+    }
+
+    /**
+     * @notice Emitted when the Hyperlane domain to Circle domain mapping is updated.
+     * @param hyperlaneDomain The Hyperlane domain.
+     * @param circleDomain The Circle domain.
+     */
+    event DomainAdded(uint32 indexed hyperlaneDomain, uint32 circleDomain);
+
+    /// @notice Hyperlane domain => Circle domain.
+    /// ATM, known Circle domains are Ethereum = 0, Avalanche = 1, Optimism = 2, Arbitrum = 3.
+    /// Note this could result in ambiguity between the Circle domain being
+    /// Ethereum or unknown.
+    mapping(uint32 => uint32) public hyperlaneDomainToCircleDomain;
+
+    /**
+     * @notice Adds a new mapping between a Hyperlane domain and a Circle domain.
+     * @param _hyperlaneDomain The Hyperlane domain.
+     * @param _circleDomain The Circle domain.
+     */
+    function addDomain(
+        uint32 _hyperlaneDomain,
+        uint32 _circleDomain
+    ) external onlyOwner {
+        hyperlaneDomainToCircleDomain[_hyperlaneDomain] = _circleDomain;
+
+        emit DomainAdded(_hyperlaneDomain, _circleDomain);
+    }
+
+    function hookType() external pure override returns (uint8) {
+        return uint8(IPostDispatchHook.Types.CCTP);
+    }
+
+    /// @inheritdoc AbstractPostDispatchHook
+    function _quoteDispatch(
+        bytes calldata metadata,
+        bytes calldata message
+    ) internal view override returns (uint256 fee) {
+        return igp.quoteDispatch(metadata, message);
+    }
+
+    /// @inheritdoc AbstractPostDispatchHook
+    function _postDispatch(
+        bytes calldata metadata,
+        bytes calldata message
+    ) internal override {
+        // Wrapped token is encoded into the metadata, see TokenBridgeCctp
+        address token = address(
+            bytes20(TokenMessage.metadata(message.body())[0:20])
+        );
+        uint256 amount = TokenMessage.amount(message.body());
+
+        IERC20(token).approve(address(tokenMessenger), amount);
```

Comment:
should we infinite approve this in the constructor? tokens should basically never be held by this hook